app/upgrades/v17/convert.go

```diff
@@ -0,0 +1,145 @@
+	accountKeeper.IterateAccounts(ctx, func(account authtypes.AccountI) (stop bool) {
+		cosmosAddress := account.GetAddress()
+		ethAddress := common.BytesToAddress(cosmosAddress.Bytes())
+		ethHexAddr := ethAddress.String()
+
+		balance, res, err := WithdrawWEVMOS(ctx, ethAddress, wrappedAddr, erc20Keeper)
+
+		var bs string // NOTE: this is necessary so that there is no panic if balance is nil when logging
+		if balance != nil {
+			bs = balance.String()
+		}
+
+		if err != nil {
+			logger.Error(
+				"failed to withdraw WEVMOS",
+				"account", ethHexAddr,
+				"balance", bs,
+				"error", err.Error(),
+			)
+		} else if res != nil && res.VmError != "" {
+			logger.Error(
+				"withdraw WEVMOS reverted",
+				"account", ethHexAddr,
+				"balance", bs,
+				"vm-error", res.VmError,
+			)
+		}
+
+		erc20Keeper.IterateTokenPairs(ctx, func(tokenPair erc20types.TokenPair) bool {
+			// NOTE: here we check if the token pair contains an IBC coin. For now we only want to convert those.
+			if !tokenPair.IsNativeCoin() {
+				return false
+			}
+
+			contract := tokenPair.GetERC20Contract()
+			if err := ConvertERC20Token(ctx, ethAddress, contract, cosmosAddress, erc20Keeper); err != nil {
+				logger.Error(
+					"failed to convert ERC20 to native Coin",
+					"account", ethHexAddr,
+					"erc20", contract.String(),
+					"balance", balance.String(),
+					"error", err.Error(),
+				)
+			}
+
+			return false
+		})
+
+		return false
+	})
+
+	// NOTE: if there are tokens left in the ERC-20 module account
+	// we return an error because this implies that the migration of native
+	// coins to ERC-20 tokens was not fully completed.
+	erc20ModuleAccountAddress := authtypes.NewModuleAddress(erc20types.ModuleName)
+	balances := bankKeeper.GetAllBalances(ctx, erc20ModuleAccountAddress)
+	if !balances.IsZero() {
+		return fmt.Errorf("there are still tokens in the erc-20 module account: %s", balances.String())
+	}
+
+	return nil
+}
+
+// ConvertERC20Token converts the given ERC20 token to the native representation.
+func ConvertERC20Token(
+	ctx sdk.Context,
+	from, contract common.Address,
+	receiver sdk.AccAddress,
+	erc20Keeper erc20keeper.Keeper,
+) error {
+	balance := erc20Keeper.BalanceOf(ctx, contracts.ERC20MinterBurnerDecimalsContract.ABI, contract, from)
+	if balance == nil {
+		return fmt.Errorf("failed to get ERC20 balance (contract %q) for %s", contract.String(), from.String())
+	}
+
+	if balance.Sign() <= 0 {
+		return nil
+	}
+
+	msg := erc20types.NewMsgConvertERC20(sdk.NewIntFromBigInt(balance), receiver, contract, from)
+	_, err := erc20Keeper.ConvertERC20(sdk.WrapSDKContext(ctx), msg)
+
+	return err
+}
+
+// WithdrawWEVMOS withdraws all the WEVMOS tokens from the given account.
+func WithdrawWEVMOS(
+	ctx sdk.Context,
+	from, wevmosContract common.Address,
+	erc20Keeper erc20keeper.Keeper,
+) (*big.Int, *evmtypes.MsgEthereumTxResponse, error) {
+	balance := erc20Keeper.BalanceOf(ctx, testdata.WEVMOSContract.ABI, wevmosContract, from)
```

Comment:
Why do we use testdata?

Reply:
It's not testdata per se, the WEVMOS contract (`WEVMOS.sol`) is just stored there - in the testdata for the WEVMOS precompile. We could move it to the `contracts` directory, but that would require changes also in the WEVMOS precompile directory and should be done on a separate PR if anything.